### ISSUE TITLE
access movable of parent macro

### ISSUE THREAD

**Issue by dschick** [sardana-org/sardana#1472]
Hi, I'd like to write a macro that is hooked in the `pre-move/post-move` hook places.
The action of the hooked macro should depend on the motor(s) moved within the parent macro (parent can be a `scan` or `mv`).

So I need a specific action only in case one specific motor is moved and not if any other is moved.

What would be the way to access the names of the movables used in the parent macro?

currently I can do

```python
parent = self.getParentMacro()
```
and then I have not found an according property/method of the parent.

Thanks in advance

Daniel

**Comment by reszelaz** [comment]
Hi Daniel,
It sounds to me familiar:) I remember we discussed it in https://github.com/sardana-org/sardana/issues/891. At that time I was suggesting to go for the GScan `moveables` member cause this one is always there, even if one writes a custom scan macro using the ScanFramework. All the standard scan macros also has `motors` attribute. Since now we are discussing if adding the `pre-move/post-move` hook places to the `mv` macros, in case we add them, then we will need to create a new member in the `mv` macro. In this case it make sense to call it `motors` as well. Then, I would reconsider my recomendation from #891 and make the `motors` a standard member of all macros having the said hook places.

**Comment by dschick** [comment]
Thanks for your response. Indeed I am especially asking also because of the `mv` macro and the according hooks.
I am in general wondering, why the scan framework is not simply refractoring the `mv` macro at every step?
Is this a problem with the continious scans?

**Comment by reszelaz** [comment]
On the last follow-up meeting we said to add motors attribute to all built-in macros executing motion and at the same time exposing `pre-move` and `post-move` hook places. This could be already done in #1480 so the `mv` and `umv` macros have this attribute.

**Comment by marouanebj** [comment]
Hi, I add my question here.
Is it possible to access from a hook the inputed params of the scan macro.
Exemple:
I have the macro "mesh" that uses 
```python
param_def = [
        ['motor1', Type.Moveable, None, 'First motor to move'],
        ['m1_start_pos', Type.Float, None, 'Scan start position for first '
                                           'motor'],
        ['m1_final_pos', Type.Float, None, 'Scan final position for first '
                                           'motor'],
        ['m1_nr_interv', Type.Integer, None, 'Number of scan intervals'],
        ['motor2', Type.Moveable, None, 'Second motor to move'],
        ['m2_start_pos', Type.Float, None, 'Scan start position for second '
                                           'motor'],
        ['m2_final_pos', Type.Float, None, 'Scan final position for second '
                                           'motor'],
        ['m2_nr_interv', Type.Integer, None, 'Number of scan intervals'],
        ['integ_time', Type.Float, None, 'Integration time'],
        ['bidirectional', Type.Boolean, False, 'Save time by scanning '
                                               's-shaped']
    ]
```
And I create a Macro "reorgData" that I put as a hook at post-scan. My macro need the bidirectional infos of the 'mesh' Macro.

I tried to put  in "reorgData" :
```python
def run(self):
        parent_macro = self.parent_macro
        self.nr_intervs = parent_macro.nr_intervs
```
But I have error parent_macro is NoneType.
I tried with self.getParentMacro() and I have the same message.

Do you know what I'm doing wrong?
Thanks

**Comment by reszelaz** [comment]
> Is it possible to access from a hook the inputed params of the scan macro.

Yes, you can use the [`getParameters()`](https://sardana-controls.org/devel/api/api_macro.html#sardana.macroserver.macro.Macro.getParameters) method.

In case of `mesh` however it already stores value of the `bidirectional_mode` parameter in _self_.

I managed to access it in the following way:

My `reorgData` macro:

```python
from sardana.macroserver.macro import macro

@macro()
def reorgData(self):
    parent_macro = self.parent_macro
    if parent_macro.name == "mesh":
        mesh = parent_macro
        bidirectional_mode = mesh.bidirectional_mode
        self.print("bidirectional_mode = {}".format(bidirectional_mode))
```

Then I configured one [general hook](https://sardana-controls.org/users/macro_hooks.html#sardana-macros-hooks-general):

```
Door_zreszela_1 [14]: defgh reorgData post-scan
Defining general hook
reorgData
```

And I tried to run the mesh scan:

```
Door_zreszela_1 [21]: mesh mot01 0 1 1 mot02 0 1 1 0.1
Operation will be saved in /tmp/scan3.h5 (Spec from SPEC_FileRecorder)
Scan #9 started at Thu Feb  4 18:22:53 2021. It will take at least 0:00:01.531371
 #Pt No    mot01     mot02      ct01      ct02      ct03      ct04    zerod01      dt   
   0         0         0        0.1       0.2       0.3       0.4     100.377   0.352107
   1         1         0        0.1       0.2       0.3       0.4     101.578   0.889462
   2         0         1        0.1       0.2       0.3       0.4     99.1231   1.43317 
   3         1         1        0.1       0.2       0.3       0.4     101.108   1.97075 
Operation saved in /tmp/scan3.h5 (Spec)
Scan #9 ended at Thu Feb  4 18:22:55 2021, taking 0:00:02.160141. Dead time 81.5% (motion dead time 64.4%)
bidirectional_mode = False

Door_zreszela_1 [22]: mesh mot01 0 1 1 mot02 0 1 1 0.1 True
Operation will be saved in /tmp/scan3.h5 (Spec from SPEC_FileRecorder)
Scan #10 started at Thu Feb  4 18:22:59 2021. It will take at least 0:00:01.531371
 #Pt No    mot01     mot02      ct01      ct02      ct03      ct04    zerod01      dt   
   0         0         0        0.1       0.2       0.3       0.4     100.102   0.353538
   1         1         0        0.1       0.2       0.3       0.4     99.6165   0.881027
   2         1         1        0.1       0.2       0.3       0.4     102.923   1.40609 
   3         0         1        0.1       0.2       0.3       0.4     101.313   1.95078 
Operation saved in /tmp/scan3.h5 (Spec)
Scan #10 ended at Thu Feb  4 18:23:01 2021, taking 0:00:02.139932. Dead time 81.3% (motion dead time 64.3%)
bidirectional_mode = True
```

**Comment by marouanebj** [comment]
I did some test and yes It works for me too when I was using spock.
But, sorry I missed to give an information, I was using the sequencer I did not realize it could be different and with the sequencer it does not work:
 self.parent_macro is NoneType.

Is it a bug ?
Thanks

**Comment by reszelaz** [comment]
@marouanebj, yes you found a bug. I already propose #1497 which should fix this issue. It would be great if you could test that PR and give us feedback if it also works for you.

**Comment by dschick** [comment]
as the initial issue is solved with #1480 I am closing this one.
